source/ShipEffectsShader.cpp

```diff
@@ -0,0 +1,370 @@
+/* ShipEffectsShader.cpp
+Copyright (c) 2014-2023 by Michael Zahniser & Daniel Yoon
+
+Endless Sky is free software: you can redistribute it and/or modify it under the
+terms of the GNU General Public License as published by the Free Software
+Foundation, either version 3 of the License, or (at your option) any later version.
+
+Endless Sky is distributed in the hope that it will be useful, but WITHOUT ANY
+WARRANTY; without even the implied warranty of MERCHANTABILITY or FITNESS FOR A
+PARTICULAR PURPOSE. See the GNU General Public License for more details.
+
+You should have received a copy of the GNU General Public License along with
+this program. If not, see <https://www.gnu.org/licenses/>.
+*/
+
+#include "ShipEffectsShader.h"
+
+#include "Color.h"
+#include "GameData.h"
+#include "Government.h"
+#include "Logger.h"
+#include "Messages.h"
+#include "Point.h"
+#include "Preferences.h"
+#include "Screen.h"
+#include "Shader.h"
+#include "Ship.h"
+#include "Sprite.h"
+
+#include <algorithm>
+#include <bit>
+#include <sstream>
+#include <vector>
+
+#ifdef ES_GLES
+// ES_GLES always uses the shader, not this, so use a dummy value to compile.
+// (the correct value is usually 0x8E46, so don't use that)
+#define GL_TEXTURE_SWIZZLE_RGBA 0xBEEF
+#endif
+
+using namespace std;
+
+namespace {
+	Shader shader;
+	GLint scaleI;
+	GLint frameI;
+	GLint frameCountI;
+	GLint positionI;
+	GLint transformI;
+	GLint blurI;
+	GLint clipI;
+
+	GLint recentHitsCountI;
+	GLint recentDamageI;
+	GLint recentHitsI;
+	GLint shieldColorI;
+	GLint ratioI;
+	GLint sizeI;
+
+	GLint fastI;
+
+	GLuint vao;
+	GLuint vbo;
+}
+
+Point ShipEffectsShader::center = Point();
+
+// Initialize the shaders.
+void ShipEffectsShader::Init()
+{
+
+	static const char *vertexCode =
+		"// vertex sprite shader\n"
+		"precision mediump float;\n"
+		"uniform vec2 scale;\n"
+		"uniform vec2 position;\n"
+		"uniform mat2 transform;\n"
+		"uniform vec2 blur;\n"
+		"uniform float clip;\n"
+
+		"in vec2 vert;\n"
+		"out vec2 fragTexCoord;\n"
+		"out vec2 shrinkby;\n"
+
+		"void main() {\n"
+		"  vec2 blurOff = 2.f * vec2(vert.x * abs(blur.x), vert.y * abs(blur.y));\n"
+		"  gl_Position = vec4((transform * (vert + blurOff) + position) * scale, 0, 1);\n"
+		"  vec2 texCoord = vert + vec2(.5, .5);\n"
+		"  shrinkby = scale;\n"
+		"  fragTexCoord = vec2(texCoord.x, min(clip, texCoord.y)) + blurOff;\n"
+		"}\n";
+
+	static const char *fragmentCode =
+		"// fragment sprite shader\n"
+		"precision mediump float;\n"
+#ifdef ES_GLES
+		"precision mediump sampler2DArray;\n"
+#endif
+		"uniform sampler2DArray tex;\n"
+		"uniform float frame;\n"
+		"uniform float frameCount;\n"
+		"uniform vec2 blur;\n"
+		"const int range = 5;\n"
+
+		"uniform vec2 recentHits[64];\n"
+		"uniform float recentDamage[64];\n"
+		"uniform int recentHitCount;\n"
+		"uniform vec4 shieldColor;\n"
+		"uniform float ratio;\n"
+		"uniform float size;\n"
```

Comment:
```suggestion
		"uniform float size = 80;\n"
```

Instead of `item.size = 80` (from my earlier feedback) you could update the shader to hard-code it

Reply:
+1